gquant/dataframe_flow/_node_flow.py

```diff
@@ -420,39 +420,16 @@ def __check_dly_processing_prereq(self, inputs):
         '''
         # check if dask future or delayed
         use_delayed = False
-        in_types = {}
-        for iport, ival in inputs.items():
-            itype = type(ival)
-            in_types[iport] = itype
+        for _, ival in inputs.items():
             if isinstance(ival, DaskDataFrame):
                 use_delayed = True
                 break
-
-        # if use_delayed:
-        #     warn_msg = \
-        #         'Node "{}" iport "{}" is of type "{}" and it '\
-        #         'should be dask_cudf.DataFrame. Ignoring '\
-        #         '"delayed_process" setting.'
-        #     for iport, itype in in_types.items():
-        #         if itype not in (dask_cudf.DataFrame,):
-        #             warnings.warn(warn_msg.format(self.uid, iport, itype))
-        #             use_delayed = False
-        #
-        # if use_delayed:
-        #     warn_msg = \
-        #         'Node "{}" oport "{}" is of type "{}" and it '\
-        #         'should be cudf.DataFrame or dask_cudf.DataFrame. Ignoring '\
-        #         '"delayed_process" setting.'
-        #     for oport, oport_spec in \
-        #             self._get_output_ports(full_port_spec=True).items():
-        #         otype = oport_spec.get('type', [])
-        #         if not isinstance(otype, list):
-        #             otype = [otype]
-        #         if dask_cudf.DataFrame not in otype and \
-        #                 cudf.DataFrame not in otype:
-        #             warnings.warn(warn_msg.format(self.uid, oport, otype))
-        #             use_delayed = False
-
+        if not use_delayed:
+            warn_msg = \
+                'None of the Node "{}" inputs '\
+                'is cudf.DataFrame or dask_cudf.DataFrame. Ignoring '\
+                '"delayed_process" setting.'.format(self.uid)
+            warnings.warn(warn_msg)
```

Comment:
While testing this, I realized this change is causing spurious warnings even when the intention is not to use automated delayed processing. This needs to be a bit smarter. I modified the code to check for dask collection property and dask futures:
https://docs.dask.org/en/latest/custom-collections.html#checking-if-an-object-is-a-dask-collection

```
# at the top import
from dask.base import is_dask_collection
from dask.distributed import Future

    def __check_dly_processing_prereq(self, inputs):
        '''At least one input must be a dask DataFrame type. Output types must
        be specified as cudf.DataFrame or dask_cudf.DataFrame. (Functionality
        could also be extended to support dask dataframe of pandas, but
        currently only cudf/dask_cudf dataframes are supported.)
        '''
        # check if dask future or delayed
        ivals = inputs.values()
        if not any((is_dask_collection(iv) for iv in ivals)) and \
                not any((isinstance(iv, Future) for iv in ivals)):
            # None of the inputs are Delayed or Futures so no intention of
            # using delayed processing. Return False and avoid printing
            # non-applicable warning.
            return False

        use_delayed = False
        for ival in ivals:
            if isinstance(ival, DaskDataFrame):
                use_delayed = True
                break

        # NOTE: Currently only support delayed processing when one of the
        #     inputs is a dask_cudf.DataFrame. In the future might generalize
        #     to support dask processing of other delayed/future type inputs.
        if not use_delayed:
            warn_msg = \
                'None of the Node "{}" inputs '\
                'is a dask_cudf.DataFrame. Ignoring '\
                '"delayed_process" setting.'.format(self.uid)
            warnings.warn(warn_msg)

        return use_delayed
```

I'm using `is_dask_collection` because it's a general way to check for dask delayed type or collection type. Example:
```
In [6]: 
   ...: def ff(a):
   ...:     return 2*a
   ...: 

In [8]: dly_ff = dask.delayed(ff)

In [9]: dly_ffa = dask.delayed(ff)(2)

In [10]: whos
Variable             Type           Data/Info
---------------------------------------------
cudf                 module         <module 'cudf' from '/hom<...>ckages/cudf/__init__.py'>
dask                 module         <module 'dask' from '/hom<...>ckages/dask/__init__.py'>
dask_cudf            module         <module 'dask_cudf' from <...>s/dask_cudf/__init__.py'>
ddf                  DataFrame      <dask_cudf.DataFrame | 2 tasks | 2 npartitions>
df                   DataFrame           a   b   c\n0    0  1<...>8   1  18\n19  19   0  19
dly_ff               DelayedLeaf    Delayed('ff-249e7176-0669<...>-4e81-9630-1d3659c08d40')
dly_ffa              Delayed        Delayed('ff-da4e05e4-b40b<...>-4a5f-925f-8ba536525ee0')
ff                   function       <function ff at 0x7ff973f47b00>
is_dask_collection   function       <function is_dask_collection at 0x7ff973eeb5f0>

In [11]: is_dask_collection(dly_ff)
Out[11]: True

In [12]: is_dask_collection(ddf)
Out[12]: True

In [13]: is_dask_collection(df)
Out[13]: False
```

Effectively `is_dask_collection` indicates whether something is delayed like.

Reply:
I changed this code. It seems that "warning" will never happen right?

Reply:
Currently in `__call__` the check is done as follows:
```
            if not self.delayed_process:
                output_df = self.decorate_process()(inputs)
            else:
                use_delayed = self.__check_dly_processing_prereq(inputs)
                if use_delayed:
                    output_df = self.__delayed_call(inputs)
                else:
                    output_df = self.decorate_process()(inputs)
```
It could be that a node has `self.delayed_process` set to true, but non of its inputs are delayed or dask collection or future, so no intention to use `__delayed_call`. The intention is implied if at least one input is delayed or a dask collection or future. This statement checks this implied intention:
```
        if not any((is_dask_collection(iv) for iv in ivals)) and \
                not any((isinstance(iv, Future) for iv in ivals)):
            # None of the inputs are Delayed or Futures so no intention of
            # using delayed processing. Return False and avoid printing
            # non-applicable warning.
            return False
```
If delayed processing is not required i.e. implied, then return false and don't print a warning. If one of the inputs can be processed in a delayed manner then we check that at least one of those inputs is a `DaskDataFrame`. Otherwise we cannot use `__delayed_call`, because we need to call `to_delayed` from one or more of the inputs. 

When one of the inputs is a Dask Delayed or a Dask Future, but none are a `DaskDataFrame` then we need to print a warning because in this situation we currently don't support automatic delay processing. The end-user should be aware in this case that their `self.delayed_process` is not valid for this situation.

Reply:
It makes sense. Currently, we don't have any case that `one of the inputs is a Dask Delayed or a Dask Future, but none are a DaskDataFrame`. But it is a good check to have.